squint.c

```diff
@@ -0,0 +1,1611 @@
+      info = *rInfo;
+      if (info & RI_hasD) {
+         // if not a memory op or a memory read, disable dest reg
+         if ((info & RI_memOp) == 0 || (info & RI_memRW)) {
+            info ^= (info & RI_RnDest) ? RI_RnAct : RI_RdAct;
+         }
+      }
+      if ((info & RI_RmAct) && (info & RI_Rm) == reg) {
+         retVal = rInfo;
+         break;
+      }
+      if (info & RI_RsAct && ((info & RI_Rs)>>8) == reg) {
+         retVal = rInfo;
+         break;
+      }
+      if (info & RI_RdAct && ((info & RI_Rd)>>12) == reg) {
+         retVal = rInfo;
+         break;
+      }
+      if (info & RI_RnAct && ((info & RI_Rn)>>16) == reg) {
+         retVal = rInfo;
+         break;
+      }
+      rInfo += direction;
+   }
+
+   return retVal;
+}
+
+/**********************************************************/
+/********** peephole optimization funcs *******************/
+/**********************************************************/
+
+void apply_peepholes1(int *funcBegin, int *funcEnd)
+{
+   int *scan;
+   int *scanp1, *scanp2, *scanp3;
+   for (scan = funcBegin; scan < funcEnd; ++scan) {
+      scan = skip_nop(scan, 1);
+      /* one instruction peepholes */
+      if (*scan == 0xe2400000 || /* sub r0, r0, #0 */
+          *scan == 0xe2800000) { /* add r0, r0, #0 */
+         *scan = NOP;
+      }
+      else {    /* two instruction peephole */
+
+         scanp1 = active_inst(scan, 1);
+
+         if ((*scan & 0xfff00000) == 0xe2800000 || // add rI, rS, #X
+             (*scan & 0xfff00000) == 0xe2400000) { // sub rI, rS, #X
+            int rS = (*scan >> 16) & 0xf; // Source, RI_Rn
+            int rI = (*scan >> 12) & 0xf; // Index,  RI_Rd
+            if ((*scanp1 & 0xff3000ff) == 0xe5100000 &&
+                ((*scanp1 >> 16) & 0xf) == rI) { // ldr[b] rX, [rI]
+
+               *scanp1 = (*scanp1 & 0xff70ff00) | (*scan & 0xff) |
+                         (((*scan & 0xfff00000) == 0xe2800000) ? (1<<23) : 0)
+                         | (rS << 16); // ldr[b] rX, [rS, #X]
+               *scan = NOP;
+            }
+         }
+      }
+   }
+
+   funcEnd -= 2;
+   for (scan = funcBegin; scan < funcEnd; ++scan) {
+      scan = skip_nop(scan, 1);
+      scanp1 = active_inst(scan,   1);
+      scanp2 = active_inst(scanp1, 1);
+
+      /* change register to immediate if possible */
+      if (*scan   == 0xe52d0004 && /* push {r0} */
+          *scanp2 == 0xe49d1004 && /* pop  {r1} */
+          (*scanp1 & 0xffffff00) == 0xe3a00000) { /* mov r0, #X */
+         scanp3 = active_inst(scanp2, 1);
+         if ((*scanp3 & 0xfe1fffff) == 0xe0010000 &&
+             (*scanp3>>23 & 0x3) != 0x2) { /* exclude comparisons */
+            *scanp3 = (*scanp3 ^ 1<<16) | 1<<25 | (*scanp1 & 0xff);
+            *scan   = NOP;
+            *scanp1 = NOP;
+            *scanp2 = NOP;
+            scan = scanp3;
+         }
+         else if (*scanp3 == 0xe1a00051) { /* asr r0, r1, r0 */
+            *scanp3 = 0xe1a00040 | (*scanp1 & 0x1f) << 7;
```

Comment:
`(*scanp1 & 0x1f)` will lead to different output (assuming asm-to-asm optimization).

In the sequence below, if `r1` is -2 initially, `r0` should be -1 after the execution according to ARM reference manual:

```
push {r0}
mov r0, #32 
pop {r1}
asr r0, r1, r0
```

However, the optimized sequence will only arithmetic shift right by 0 and `r0` will become `-2`:


```
asr r0, r1, #0
```

May be check whether the constant value is with in 0 <= shift_amount <= 31 before applying this optimization?

Reply:
Good eye.  I will verify.